### PR TITLE
ENH: Support sorting frames by a combo of columns and index levels (GH 14353)

### LINKED ISSUE #14353: ENH: Support sorting DataFrames by a combination of columns and index levels
#### Background

During the review of @jreback's PR last year to cleanup the sorting API (#10726) there was some discussion of how the DataFrame API could eventually support sorting by a combination of columns and index levels.  I'm interested in working on implementing this soon and would like to continue the discussion of where this should fit into the DataFrame sorting API.

In https://github.com/pydata/pandas/pull/10726#issuecomment-128066486 @jorisvandenbossche made the following suggestion

> If we want to add this enhancement to simultaneously specify to sort on index levels and columns (the 5d option of above), then the question is: where do we add this functionality and how? In sorted, sort_index or both? I would then lean towards saying: only add it in sorted, where the by keyword can also denote a index level name.

This approach makes good sense to me. Each object passed to the `by` keyword of `sort_values` (referred to as `sorted` in the quote above) could refer to either a column or an index level.  For backwards compatibility, column references would take precedence. And my assumption is that we would want to preserve the index when sorting by a combination of columns and index levels this way.

This proposal is the sorting analog of the groupby proposal in #5677 (which I will be working on soon)

### PR DESCRIPTION
**Pull request by jonmmease**
This PR implements the changes proposed in #14353. @jorisvandenbossche 

- [x] closes #14353
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [x] Rework to use utility methods introduced in #17484 after it is merged

### CODE REVIEW

File: doc/source/basics.rst
Comment (gfyoung) [inline doc/source/basics.rst]: I know this word was in the original doc, but I'm a little uneasy about judgmental words like "obvious."  I think we can just remove it.

File: doc/source/basics.rst
Comment (gfyoung) [inline doc/source/basics.rst]: Not sure you need this version-added tag here since you add it below.

File: pandas/core/generic.py
Comment (gfyoung) [inline pandas/core/generic.py]: *Not blocking*: My OCD really wants you to not "orphan" the "levels" word at the end of this sentence :smile: .  However, that's largely a me problem, but if you do happen to come up with a wording that doesn't do this orphaning, by all means go for it.

File: doc/source/basics.rst
Comment (jmmease) [reply doc/source/basics.rst]: Agreed, I've reworded it.

File: doc/source/basics.rst
Comment (jmmease) [reply doc/source/basics.rst]: I left this versionadded tag and removed the one below in the note. This will help people quickly realize that the feature is new and it should be clear that the note below wouldn't apply without the feature.

File: pandas/core/generic.py
Comment (jmmease) [reply pandas/core/generic.py]: Heh, now it bothers me too! I reworked it to fit on one line. "Name or list of names matching axis items or index levels."

**Comment by jreback** [comment]
@jmmease this looks like an older version of #17484

**Comment by jmmease** [comment]
@jreback This PR was for the sort_values col&index change. It had some shared utility code with #17484  (the merge col&index change) so i was holding off on updating it until that one was merged.  

Do you want to keep this PR closed until #17484 is merged and I make the updates?

**Comment by jreback** [comment]
np can rebase and evaluate after #17484

**Comment by jmmease** [comment]
Ok, I just pushed a bunch of updates:
- Converted to use the new `_get_label_or_level_values` utility method introduced in #17484
- Moved tests to a new file, updated to use fixtures/parametrize, and added `axis=1` test cases
- I reviewed the `sort_values` documentation in the docstrings and in `basics.rst` and made small language updates throughout.  I hope these changes make things a bit clearer and more consistent.
- I added a full whatsnew section for 0.22.0

@jreback @jorisvandenbossche @gfyoung @TomAugspurger

**Comment by jmmease** [comment]
Just following up on this. 

@gfyoung if you have time to look this over at some point, I'd appreciate your feedback on these documentation/docstring updates.

@jreback does this approach look alright to you?

Thanks all

**Comment by jreback** [comment]
@jmmease lgtm. @TomAugspurger @jorisvandenbossche if you'd have a look.

File: pandas/core/frame.py
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: I find "off-axis labels" very complex. I mean I know what you mean and theoretically it is correct, but I think novice users will not understand this (the previous "axis items" was also not that good). 

I am still thinking of a better wording (the annoying thing is that you cannot just say "column labels or index level names" which is probably the common case how this is used)

File: pandas/tests/frame/test_sort_values_level_as_str.py
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_sort_values_level_as_str.py]: Can you see if you can get this working without `check_stacklevel=False` ?

File: pandas/tests/frame/test_sorting.py
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_sorting.py]: I think for this case the old error message was more informative, so it would be nice if we could keep a separate message for the case of incomplete key for MI compared to just non-unique key

File: pandas/core/frame.py
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: See comment below as well, so it would be nice to somehow keep this message

**Review by jorisvandenbossche — COMMENTED** [review]
@jmmease Added some comments!

File: pandas/tests/frame/test_sort_values_level_as_str.py
Comment (jmmease) [reply pandas/tests/frame/test_sort_values_level_as_str.py]: Sure.  Could you offer any insight into what `check_stacklevel` is actually checking? I've had trouble understanding that @jorisvandenbossche

File: doc/source/basics.rst
Comment (jmmease) [reply doc/source/basics.rst]: Maybe this isn't correct. But I was trying to be consistent in using `level` to refer to the names of the indexes or columns (`df.index.names`, `df.columns.names`) and `label` to refer to the values referenced by the levels (`df.index.get_level_values(lv1)`, `df.columns.get_level_values(lv1)`). Is that in line with accepted vocabulary? @jorisvandenbossche

Comment (jorisvandenbossche) [reply doc/source/basics.rst]: Personally I think "index level" is the equivalent of "column", and from there it would follow that the equivalent of "column values" is "index labels of one level / index level labels / index level values" (but none of those sound that good) and the equivalent of "column names/labels" would then be "index level name". 
But it's not that we do that very consistently currently, or that this is necessarily agreed upon vocabulary. But I think it is in line more or less with what you say?

The main reason I commented it here, is because the next line uses "column values", and as you say "label to refer to the values" then to be consistent it would be "index labels". 
And a second reason I commented is that I think "index level" is more complicated, as then a user needs to be familiar with the concept of MultiIndexes, while for this section about sorting, that is not needed.

File: pandas/tests/frame/test_sort_values_level_as_str.py
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_sort_values_level_as_str.py]: See https://github.com/pandas-dev/pandas/issues/9584 and https://github.com/pandas-dev/pandas/pull/10676. 
The `stacklevel` is specified in the FutureWarning in `_check_label_or_level_ambiguity`, and is currenlty set at `2`, which is not correct. This should be higher, depending on how many call steps there are between `sort_values` and calling that function. 
The main problem here will probably be that  `_check_label_or_level_ambiguity` is used in multiple places, so the needed stacklevel might differ for the different usages.

File: doc/source/basics.rst
Comment (TomAugspurger) [inline doc/source/basics.rst]: Same issue here with "index levels". Does changing it to "index names" make sense here, since that's what you're referring to?

Side-question, does this work on a regular (non-multiindex) index?

Comment (TomAugspurger) [reply doc/source/basics.rst]: I see you use "index level names" in the whatsnew. That seems like a good option, if you could use it here too.

File: pandas/tests/frame/test_sort_values_level_as_str.py
Comment (jmmease) [reply pandas/tests/frame/test_sort_values_level_as_str.py]: Thanks, that's helpful

File: pandas/core/frame.py
Comment (TomAugspurger) [inline pandas/core/frame.py]: This could be a `..versionmodified:`

    .. versionmodified:: 0.22.0
       Allow specifying index or column level names.

File: pandas/core/frame.py
Comment (TomAugspurger) [reply pandas/core/frame.py]: How about "Name or list of names to sort by"? And then things are clarified below.

File: pandas/tests/frame/test_sort_values_level_as_str.py
Comment (TomAugspurger) [reply pandas/tests/frame/test_sort_values_level_as_str.py]: The goal is to report the users's line of code that triggered the warning. In our code, when you emit the `FutureWarning` there's an optional `stacklevel` parameter that's controls which line of code the error message is reported with. It's supposed to be the number of function calls from user code to our method emitting the warning.

It's not always possible to get that correct, if you have multiple pandas methods calling the method emitting the warning.

File: doc/source/basics.rst
Comment (jmmease) [reply doc/source/basics.rst]: Thanks for clarifying.  After looking at it again, I do agree that "index labels" is the better parallel to "columns values".

File: doc/source/basics.rst
Comment (jmmease) [reply doc/source/basics.rst]: I split this into two sentences so that a `Series` is sorted by "values" and a `DataFrame` is sorted by "column or row values".

File: doc/source/basics.rst
Comment (jmmease) [reply doc/source/basics.rst]: Sounds good

File: pandas/core/frame.py
Comment (jmmease) [reply pandas/core/frame.py]: Sounds good. Going with "Name or list of names to sort by"

File: pandas/tests/frame/test_sort_values_level_as_str.py
Comment (jmmease) [reply pandas/tests/frame/test_sort_values_level_as_str.py]: Ok, I was able to remove `check_stacklevel=False` for these tests, and for the related tests for `groupby` and `merge`.  I added a `stacklevel` parameter to `_check_label_or_level_ambiguity` and `_get_label_or_level_values` that behaves like the `stacklevel` parameter of `warnings.warn` itself (`stacklevel=1` blames the caller, `stacklevel=2` blames the callers parent, etc.).

Does this look like a good way to handle it? @TomAugspurger @jorisvandenbossche

File: pandas/tests/frame/test_sorting.py
Comment (jmmease) [reply pandas/tests/frame/test_sorting.py]: Ok, sure. The new error message will read as follows (for the `MultiIndex` case):
```
ValueError: The column label 'a' is not unique.
For a multi-index, the label must be a tuple with elements corresponding to each level.
```
Does that sound clear? @jorisvandenbossche

File: doc/source/basics.rst
Comment (jmmease) [reply doc/source/basics.rst]: Changed to "index level names". 

And yes, this does work for both the `Index` and `MultIndex` cases.
@TomAugspurger

File: pandas/core/frame.py
Comment (jmmease) [reply pandas/core/frame.py]: done

**Comment by jmmease** [comment]
Thanks for the thorough comments @jorisvandenbossche and @TomAugspurger 

I just pushed an update that I believe address all of them.  Let me know if I missed anything or if you see any other updates you'd like before this is merged.

**Comment by jmmease** [comment]
@TomAugspurger @jorisvandenbossche Did these latest updates address your comments? Are there any other changes you'd like to see before this is merged? Thanks!

cc @jreback

**Comment by jmmease** [comment]
Friendly ping
@TomAugspurger @jorisvandenbossche @jreback

**Comment by jmmease** [comment]
@jreback @jorisvandenbossche @TomAugspurger  I noticed the milestone tag for this was moved to 0.23.0.  I just moved the whatsnew to this version. Are we in a freeze for merging things into 0.23.0 until 0.22.0 is released?

**Comment by jreback** [comment]
@jmmease 0.22 is a special release coming out today. so all previous 0.22 is now 0.23.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@jmmease just a couple of minor comments. note that windows CI is currently failing.

**Comment by jmmease** [comment]
Thanks @jreback . If I'm not mistaken, It looks like the comments you just made ("blank line", and "issue number") are against an older version of the PR. The new tests were parametrized and moved to `test_sort_values_level_as_str.py`.

**Comment by jmmease** [comment]
@jreback The AppVeyor build is passing now, but I'm getting a "The job exceeded the maximum time limit for jobs, and has been terminated." error for most of the Travis builds.

**Comment by jreback** [comment]
yeah travis just started failing
not sure why

**Comment by jreback** [comment]
rebase and push again, fixed some hanging by Travis CI

**Comment by jmmease** [comment]
Done. Thanks @jreback . Happy New Year BTW!

**Comment by jreback** [comment]
happy new year. will have a final look, but this lgtm. @TomAugspurger

**Review by TomAugspurger — APPROVED** [review]

**Comment by jmmease** [comment]
Thanks for taking a look @TomAugspurger 
Anything else @jreback ?

**Comment by jreback** [comment]
thanks @jmmease really nice patch! keep em coming!

**Comment by jmmease** [comment]
Awesome, thanks for all your time on this @jreback @TomAugspurger, @jorisvandenbossche, and @gfyoung !

**Pull request merged**
